runtime/lua/vim/diagnostic.lua

```diff
@@ -1348,16 +1334,10 @@ end
 
 M.handlers.signs = {
   show = function(namespace, bufnr, diagnostics, opts)
-    vim.validate({
-      namespace = { namespace, 'n' },
-      bufnr = { bufnr, 'n' },
-      diagnostics = {
-        diagnostics,
-        vim.islist,
-        'a list of diagnostics',
```

Comment:
This is a minor regression. Why is performance important for this function?

Similar question for other cases.

If we can make the faster version support caller-defined validators, that could be interesting.

Reply:
It's a regression, but not one I think is meaningful. Checking for a list isn't particular valuable, and the functions will _work_ with lists with holes, it'll just behave similarly to how `ipairs` does.

I can add it back if you like, but I think simply checking for `table` is good enough.

For other complex validators, I've left them in.

Reply:
I've backtracked on this because of this test failure:

```lua
      matches(
        'expected a list of diagnostics',
        pcall_err(exec_lua, [[vim.diagnostic.set(1, 0, {lnum = 1, col = 2})]])
      )
```

Specifically:
```lua
vim.diagnostic.set(1, 0, {lnum = 1, col = 2}) -- fails
vim.diagnostic.set(1, 0, {{lnum = 1, col = 2}}) -- passes
```

which I think is meaningful enough.